### PR TITLE
New: Docs - UI States [ED-4628]

### PR DESCRIPTION
**Pull request by StyleShit**

### CODE REVIEW

File: docs/core/common/assets/js/api/core/ui-states.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+		}
+ 
+		onChange( oldValue, newValue ) {
+			// Callback that runs on state change.
+			console.log( { oldValue, newValue } );
+		}
+	}
+	```
+
+  > Legend
```
Comment (nisan250) [inline docs/core/common/assets/js/api/core/ui-states.md:98]: What does it mean?

File: docs/core/common/assets/js/api/core/ui-states.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+	// OR
+
+	elementor.$previewContents[ 0 ].addEventListener( 'e-ui-state:document/direction-mode', ( e ) => {
+		const { oldValue, newValue } = e.detail;
+		
+		console.log( 'Direction mode changed', { oldValue, newValue } );
+	} );
+	```
+
+### [Back](../readme.md)
```
Comment (nisan250) [inline docs/core/common/assets/js/api/core/ui-states.md:178]: Not sure if it should be added, but, don't you think maybe to write an explanation of why this feature has been added (reasons and such) and what benefits does it provide?

File: docs/core/common/assets/js/api/core/ui-states.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+|----------------------------|----------------------------------------------------------|-----------|---------------------------------------------------------------------
+| `$e.uiStates.register()`   |`{UiStateBase}` *instance*   								| `{void}`  | Register a new UI state.
+| `$e.uiStates.getAll()`     |                                                          | `{Object}`| Get all existing UI states with their options.
+| `$e.uiStates.get()`        | `{String}` *state*                                       | `{Object}`| Get the state value, or return all of them if no `state` is set.
+| `$e.uiStates.set()`        | `{String}` *state*, `{String}` *value*                   | `{void}`  | Set the current state value and trigger its callbacks & events.
+| `$e.uiStates.remove()`     | `{String}` *state*                                       | `{void}`  | Remove a state.
+| `$e.uiStates.getCurrent()` | `{String}` *state*                                       | `{string}`| Get the current state value.
+* **Notes**:
+	- Each UI state has its own options & callback for each option.
```
Comment (nisan250) [inline docs/core/common/assets/js/api/core/ui-states.md:15]: Each UI state has its options & callback for each option.

File: docs/core/common/assets/js/api/core/ui-states.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+	- When a UI state is being changed, it:
+		- Adds a CSS class to the scope elements ( `e-ui-state--${ stateID }__${ value }` ) - Slashes are being replaced by hyphens ( e.g. `document/direction-mode` will become `document-direction-mode` ).
+		- Dispatches a custom event to the scope elements ( `e-ui-state:${ stateID }`, e.g. `e-ui-state:document/direction-mode` ) with `oldValue` and `newValue` under `e.detail`.
+	- Setting an invalid option to a state will throw an error.
+
+## Guidelines, Conventions & Files Structure:
+* Each UI state is owned by a [component](./components.md#guidelines-conventions--files-structure).
+* Each [component](./components.md#guidelines-conventions--files-structure) can override the `defaultUiStates()` method to return a set of UI states of its own.
+* The UI states are imported via built-in method called `importUiStates()`.
```
Comment (nisan250) [inline docs/core/common/assets/js/api/core/ui-states.md:24]: The UI states are imported via a built-in

File: docs/core/common/assets/js/api/core/ui-states.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	* All the UI states should be exported in one index file:
+	    ```javascript
+		// ui-states/index.js
+		export { DirectionMode } from './direction-mode';
+		export { ColorPickingMode } from './color-picking-mode';
+		```
+	  You can have as many indexes as you wish in every hierarchy under `component/ui-states/some/long/path` in order to organize your code, but the requirement is to have a single index file
```
Comment (nisan250) [inline docs/core/common/assets/js/api/core/ui-states.md:31]: You can have as many indexes as you wish in every hierarchy under `component/ui-states/some/long/path` to organize

**Review by nisan250 — CHANGES_REQUESTED** [review]

File: docs/core/common/assets/js/api/core/ui-states.md (hunk 1 above)
Comment (StyleShit) [reply docs/core/common/assets/js/api/core/ui-states.md:98]: https://translate.google.com/?sl=iw&tl=en&text=%D7%9E%D6%B4%D7%A7%D7%A8%D6%B8%D7%90&op=translate

See also @iNewLegend's docs, it's the same as there.

File: docs/core/common/assets/js/api/core/ui-states.md (hunk 3 above)
Comment (StyleShit) [reply docs/core/common/assets/js/api/core/ui-states.md:15]: Done.

File: docs/core/common/assets/js/api/core/ui-states.md (hunk 2 above)
Comment (StyleShit) [reply docs/core/common/assets/js/api/core/ui-states.md:178]: Good idea
Done.

**Review by nisan250 — APPROVED** [review]

File: docs/core/common/assets/js/api/core/ui-states.md [hunk 6]
```diff
@@ -0,0 +1,180 @@
+## API - `$e.uiStates`
+*  **Description**: `$e.uiStates` API is a UI state manager that allows you to create custom UI states for your components.
+   We've added this API since we needed a way to reflect the component's state in the UI itself and react accordingly
```
Comment (matipojo) [inline docs/core/common/assets/js/api/core/ui-states.md:3]: Please remove all `we` / `you` / `us` etc. in the docs.

```suggestion
   It reflects the component's state in the UI itself and allows to react accordingly
```

File: docs/core/common/assets/js/api/core/ui-states.md [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+   ( i.e. rotate icons, hide buttons, etc. ).  
+*  **Location**: *core/common/assets/js/api/core/ui-states.js*
+*  **Methods**:
+
+| Method                     | Params                                                   | Returns   | Description
+|----------------------------|----------------------------------------------------------|-----------|---------------------------------------------------------------------
+| `$e.uiStates.register()`   |`{UiStateBase}` *instance*   								| `{void}`  | Register a new UI state.
+| `$e.uiStates.getAll()`     |                                                          | `{Object}`| Get all existing UI states with their options.
+| `$e.uiStates.get()`        | `{String}` *state*                                       | `{Object}`| Get the state value, or return all of them if no `state` is set.
```
Comment (matipojo) [inline docs/core/common/assets/js/api/core/ui-states.md:12]:
```suggestion
| `$e.uiStates.get()`        | `{String}` *state* (optional)                                       | `{Object}`| Get the state value, or return all of them if no `state` is set.
```

File: docs/core/common/assets/js/api/core/ui-states.md [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+| `$e.uiStates.set()`        | `{String}` *state*, `{String}` *value*                   | `{void}`  | Set the current state value and trigger its callbacks & events.
+| `$e.uiStates.remove()`     | `{String}` *state*                                       | `{void}`  | Remove a state.
+| `$e.uiStates.getCurrent()` | `{String}` *state*                                       | `{string}`| Get the current state value.
+* **Notes**:
+	- Each UI state has options and an optional callback for each option ( a callback that will execute each time the state has changed to this value ).
```
Comment (matipojo) [inline docs/core/common/assets/js/api/core/ui-states.md:17]: Please remove all spaces between the text and the brackets, here and next.
```suggestion
	- Each UI state has options and an optional callback for each option (a callback that will execute each time the state has changed to this value).
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: docs/core/common/assets/js/api/core/ui-states.md (hunk 6 above)
Comment (StyleShit) [reply docs/core/common/assets/js/api/core/ui-states.md:3]: Done

File: docs/core/common/assets/js/api/core/ui-states.md (hunk 7 above)
Comment (StyleShit) [reply docs/core/common/assets/js/api/core/ui-states.md:12]: Done

File: docs/core/common/assets/js/api/core/ui-states.md [hunk 9]
```diff
@@ -0,0 +1,180 @@
+## API - `$e.uiStates`
+*  **Description**: `$e.uiStates` API is a UI state manager that allows you to create custom UI states for your components.
+   It reflects the component's state in the UI itself and allows you to react accordingly
```
Comment (matipojo) [inline docs/core/common/assets/js/api/core/ui-states.md:3]: Please remove the `you` across the doc.
```suggestion
   It reflects the component's state in the UI itself and allows to react accordingly
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: docs/core/common/assets/js/api/core/ui-states.md (hunk 9 above)
Comment (StyleShit) [reply docs/core/common/assets/js/api/core/ui-states.md:3]: Done

**Review by matipojo — APPROVED** [review]

**Pull request merged**